src/hotspot/share/tsan/tsanOopMap.cpp

```diff
@@ -414,81 +247,86 @@ namespace TsanOopMapImpl {
-    // overwriting. Overall, this can take N^2 steps if only one object can be
-    // moved during the array traversal; however, when we are dealing with
-    // compacting garbage collector, observation shows that the overwhelming
-    // majority of the objects move in one direction. If we sort the moves (in
-    // the ascending order if dominant direction is downwards, in the descending
-    // order otherwise), chances are we will be able to order the moves in a few
-    // traversals of the moves array.
-    {
-      moves.sort((2 * n_downward_moves > moves.length()) ? lessThan : moreThan);
-      log_debug(gc)("Tsan: sort %d objects", moves.length());
-    }
+OopStorage* TsanOopMap::oop_storage() {
+  assert(_weak_oop_storage != NULL, "sanity");
+  return _weak_oop_storage;
+}
 
-    {
-      handle_overlapping_moves(moves, min_low, max_high);
-    }
+// Called during GC by WeakProcessor.
+void TsanOopMap::notify_tsan_for_freed_and_moved_objects() {
+  if (_oop_map == nullptr) {
+    return;
   }
 
-}  // namespace TsanOopMapImpl
+  bool disjoint_regions;
+  int n_downward_moves = 0;
+  char *min_low, *max_high;
+  char *source_low = reinterpret_cast<char *>(UINTPTR_MAX);
+  char *source_high = NULL;
+  char *target_low = reinterpret_cast<char *>(UINTPTR_MAX);
+  char *target_high = NULL;
 
+  ResourceMark rm;
+  GrowableArray<TsanOopMapImpl::PendingMove> moves(MAX2((int)(_oop_map->size()), 100000));
 
-void TsanOopMap::initialize_map() {
-  TsanOopMapImpl::oop_map = new TsanOopMapImpl::TsanOopSizeMap(512);
-}
-
-void TsanOopMap::destroy() {
-  delete TsanOopMapImpl::oop_map;
-}
+  {
+    MutexLocker mu(TsanOopMap_lock, Mutex::_no_safepoint_check_flag);
+    _oop_map->collect_moved_objects_and_notify_freed(
+                                 &moves, &source_low, &source_high,
+                                 &target_low, &target_high,
+                                 &n_downward_moves);
+  }
 
-void TsanOopMap::weak_oops_do(
-    BoolObjectClosure* is_alive,
-    OopClosure* pointer_adjuster) {
-  if (!ThreadSanitizer) return;
-  assert(SafepointSynchronize::is_at_safepoint(), "must be at a safepoint");
+  // No lock is needed after this point.
+  if (moves.length() != 0) {
+    // Notify Tsan about moved objects.
+    disjoint_regions = (source_low >= target_high || source_high <= target_low);
+    min_low = MIN2(source_low, target_low);
+    max_high = MAX2(source_high, target_high);
 
-  // We're mutating oopMap, but we don't need to acquire TsanOopMap_lock:
-  // Mutation to map happens at (A) constructor (single threaded) and
-  // (B) add (in mutator threads) and (C) do_weak_oops (single-threaded).
-  // Calls between add are synchronized.
-  // Calls between add and do_weak_oops are synchronized via STW GC.
-  TsanOopMapImpl::TsanOopSizeMap::rebuild_oops_map(
-      is_alive, pointer_adjuster);
+    moves.sort((2 * n_downward_moves > moves.length()) ?
```

Comment:
Sorting is only needed before calling `handle_overlapping_moves` right? Then it should be in the else branch.

Reply:
I moved around `moves.sort` a few times for the disjoint and overlapping cases during implementing/testing, due to the `((*dst_meta)) == ((0))` check failures. 

I agree that it only needs sorting for the overlapping case. I just retested with sorting for overlapping case only. All tsan jtreg tests still pass with a release JDK binary.